partner_company_default/models/res_partner.py

```diff
@@ -1,10 +1,17 @@
 # Copyright 2023 Quartile Limited
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
 
-from odoo import fields, models
+from odoo import api, fields, models
 
 
 class ResPartner(models.Model):
     _inherit = "res.partner"
 
     company_id = fields.Many2one(default=lambda self: self.env.company)
+
+    @api.model
+    def create(self, vals):
+        # The context value is set in the create method of res.company
+        if self.env.context.get("creating_from_company"):
+            vals["company_id"] = False
+        return super(ResPartner, self).create(vals)
```

Comment:
Previous code is not multi-record creation aware, but I have tried and it's not always correct depending on when the create override in the stack is executed, so better to put company_id=False from the beginning this way:

```suggestion
    company_id = fields.Many2one(default=lambda self: self._default_company_id())

    @api.model
    def _default_company_id(self):
        """Return False for other tests or if creating a company."""
        context = self.env.context
        if (
            context.get("creating_from_company")
            or config["test_enable"]
            and not context.get("test_partner_company_default")
        ):
            return False
        return self.env.company
```

and on tests, pass the corresponding context.